File: lobby/src/main/java/xyz/openatbp/lobby/Player.java
package xyz.openatbp.lobby;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.DataOutputStream;
import java.io.IOException;
import java.net.Socket;

public class Player {

    private float pid;
    private String name;
    private String avatar;
    private String tegid;
    private boolean ready;
    private Socket conn;
    private String team;

    private static ObjectMapper objectMapper = new ObjectMapper();


    public Player(Socket conn, JsonNode playerObj){
        this.conn = conn;
        this.pid = (float)playerObj.get("auth_id").asDouble();
        this.name = playerObj.get("name").asText();
        this.avatar = "unassigned";
        this.tegid = playerObj.get("teg_id").asText();
        this.ready = false;
    }

    @Deprecated public JsonNode toObject(){ //Turns Player obj to Json Object
        ObjectNode playerObj = objectMapper.createObjectNode();
        playerObj.put("name",this.name);
        playerObj.put("player",this.pid);
        playerObj.put("teg_id",this.tegid);
        playerObj.put("avatar",this.avatar);
        playerObj.put("is_ready",this.ready);

        return playerObj;
    }

    public void setAvatar(JsonNode node){
        this.avatar = node.get("name").asText();
    } //Avatar setter

    public void setReady(){
        this.ready = true;
    } //Ready setter

    public DataOutputStream getOutputStream(){ //Gets output stream to send info to client
        try{
            return new DataOutputStream(this.conn.getOutputStream());
        }
        catch (IOException e){
            return null;
        }
    }

    public boolean isAddress(String address){ //Returns if address matches player's address
        return this.conn.getRemoteSocketAddress().toString().equals(address);
    }

    public String getUsername(){ //Returns TEGid
        return this.tegid;
    }

    public String getName() { //Returns display name
        return this.name;
    }

    public float getPid(){ //Returns authid
        return this.pid;
    }

    public String getAvatar(){ // Returns avatar
        return this.avatar;
    }

    public boolean isReady(){ // Returns ready status
        return this.ready;
    }

    public void leaveTeam(){
        this.avatar = "unassigned";
        this.ready = false;
        this.team = null;
    }
    public String getTeam(){
        return this.team;
    }

    public void setTeam(String team){
        this.team = team;
    }
}

File: lobby/src/server/player.ts
import type { Socket, AddressInfo } from 'node:net';

import { send } from './packet';
import type Queue from './queue';
import type { ResponsePayload } from './response.type';

export type PlayerObject = {
  name: string,
  auth_id: number,
  teg_id: string,
};

export default class {
  private pid: number;

  private name: string;

  private avatar: string;

  private tegid: string;

  private ready: boolean;

  private socket: Socket;

  private team: string | undefined;

  private queue: Queue | undefined;

  constructor(socket: Socket, playerOrGuestNum: PlayerObject | number) {
    this.socket = socket;
    this.avatar = 'unassigned';
    this.ready = false;
    this.team = undefined;
    this.queue = undefined;
    if (typeof playerOrGuestNum === 'number') {
      this.pid = playerOrGuestNum;
      this.name = 'Guest';
      this.tegid = 'Guest';
    } else {
      this.pid = playerOrGuestNum.auth_id;
      this.name = playerOrGuestNum.name;
      this.tegid = playerOrGuestNum.teg_id;
    }
  }

  public toObject = () => ({
    name: this.name,
    player: this.pid,
    teg_id: this.tegid,
    avatar: this.avatar,
    is_ready: this.ready,
  });

  public getPid = (): number => this.pid;

  public getName = (): string => this.name;

  public getAvatar = (): string => this.avatar;

  public getUsername = (): string => this.tegid;

  public getReady = (): boolean => this.ready;

  public getOutputStream = (): Socket => this.socket;

  public getAddress = (): string => (this.socket.address() as AddressInfo).address;

  public getTeam = (): string | undefined => this.team;

  public getQueue = (): Queue | undefined => this.queue;

  public setAvatar = (avatar: string) => { this.avatar = avatar; };

  public setReady = (ready: boolean = true) => { this.ready = ready; };

  public setTeam = (team: string) => { this.team = team; };

  public setQueue = (q: Queue) => { this.queue = q; };

  public isReady = (): boolean => this.getReady();

  public isAddress = (address: string): boolean => this.getAddress() === address;

  public leaveTeam = () => {
    this.avatar = 'unassigned';
    this.ready = false;
    this.team = undefined;
    this.queue = undefined;
  };

  public send = (payload: ResponsePayload) => send(this.socket, JSON.stringify(payload));
}
